// ionic-tooltips/esm5/lib/tooltips.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TooltipBox } from './tooltip-box.component';
import { TooltipController } from './tooltip.cotroller';
import { TooltipDirective } from './tooltip.directive';
var TooltipsModule = /** @class */ (function () {
    function TooltipsModule() {
    }
    /**
     * @return {?}
     */
    TooltipsModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: TooltipsModule,
            providers: [TooltipController],
        };
    };
    TooltipsModule.decorators = [
        { type: NgModule, args: [{
                    entryComponents: [TooltipBox],
                    declarations: [TooltipDirective, TooltipBox],
                    imports: [CommonModule],
                    exports: [TooltipDirective],
                },] }
    ];
    return TooltipsModule;
}());
export { TooltipsModule };
